File: GUILib/src/space/kiibou/event/MouseEventButton.java
package space.kiibou.event;

import processing.core.PConstants;

import java.util.HashMap;
import java.util.Map;

public enum MouseEventButton {
    LEFT(PConstants.LEFT),
    RIGHT(PConstants.RIGHT),
    CENTER(PConstants.CENTER);

    private static final Map<Integer, MouseEventButton> mapper = new HashMap<>();

    static {
        for (MouseEventButton button : MouseEventButton.values()) {
            mapper.put(button.id, button);
        }
    }

    private int id;

    MouseEventButton(int id) {
        this.id = id;
    }

    public static MouseEventButton fromProcessingEvent(processing.event.MouseEvent event) {
        return mapper.get(event.getButton());
    }
}

File: GUILib/src/space/kiibou/event/MouseEventButton.kt
package space.kiibou.event

import processing.core.PConstants
import processing.event.MouseEvent

enum class MouseEventButton(private val id: Int) {
    LEFT(PConstants.LEFT),
    RIGHT(PConstants.RIGHT),
    CENTER(PConstants.CENTER);

    companion object {
        private val mapper: Map<Int, MouseEventButton> = values().map { it.id to it }.toMap()

        fun fromProcessingEvent(event: MouseEvent): MouseEventButton {
            return mapper[event.button]
                    ?: error("Could not associate processingMouseButton:${event.button} with a MouseEventButton")
        }
    }

}